hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/ComplexAvroKeyGenerator.java

```diff
@@ -36,6 +38,9 @@ public ComplexAvroKeyGenerator(TypedProperties props) {
         .split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
     this.partitionPathFields = Arrays.stream(props.getString(KeyGeneratorOptions.PARTITIONPATH_FIELD_NAME.key())
         .split(",")).map(String::trim).filter(s -> !s.isEmpty()).collect(Collectors.toList());
+    this.indexKeyFields = props.getStringList(
```

Comment:
Also why would this be different than the record key fields? We bucket by those right, only then Can we make sure Updates to same key k goes into the same bucket?

Reply:
> Can we keep the key generator unchanged? Extract the index field during actual
> indexing? I find it odd that key generator has to know about index field. I guess you
> want to do this for efficiency?

Keep the index field is to be more efficient without deserialization

Reply:
Bucket key needs to be a subset of the record key,  so that records with the same key can go into the same bucket. e.g. bucket key is `colA` and record key is `id` + `colA`.
Generally, the record key is `id` to identified the record, and bucket key is more like an attribute to cluster the data. When using bucket index, bucket key needs to be added to record key to meet the data distribution and uniquely identify a record.

Reply:
Good suggestion, will fix